### ISSUE TITLE
[API Proposal]: AsyncConfigurationProvider

### ISSUE THREAD

**Issue by KieranDevvs** [dotnet/runtime#79193]
### Background and motivation

As it currently stands, configuration sources have no way of being asynchronous (Highlighted in #36018).
These new set of types should allow for users to implement asynchronous configuration providers without any breaking changes to the way the existing API's function.

### API Proposal

```diff
namespace Microsoft.Extensions.Configuration;

public partial class ConfigurationBuilder : IConfigurationBuilder
{
+    public IConfigurationBuilder Add(IAsyncConfigurationSource source);
}
```

```cs
namespace Microsoft.Extensions.Configuration;

public interface IAsyncConfigurationSource
{
    IAsyncConfigurationProvider Build(IConfigurationBuilder builder);
}
```

```cs
namespace Microsoft.Extensions.Configuration;

public interface IAsyncConfigurationProvider
{
    bool TryGet(string key, out string? value);
    void Set(string key, string? value);
    IChangeToken GetReloadToken();
    Task LoadAsync();
    IEnumerable<string> GetChildKeys(IEnumerable<string> earlierKeys, string? parentPath);
}
```

```cs
namespace Microsoft.Extensions.Configuration;

public abstract class AsyncConfigurationProvider : IAsyncConfigurationProvider
{
    protected IDictionary<string, string?> Data { get; set; }
    public virtual bool TryGet(string key, out string? value);
    public virtual void Set(string key, string? value);
    public virtual Task LoadAsync();
    public virtual IEnumerable<string> GetChildKeys(IEnumerable<string> earlierKeys, string? parentPath);
    public IChangeToken GetReloadToken();
    protected void OnReload()
}
```

### API Usage

```csharp
internal class Program
{
    static void Main(string[] args)
    {
        var builder = new ConfigurationBuilder();
        builder.Add(new DatabaseConfigurationSource());
    }
}

public class DatabaseConfigurationSource : IAsyncConfigurationSource
{
    public IAsyncConfigurationProvider Build(IConfigurationBuilder builder)
    {
        return new DatabaseConfigurationProvider ();
    }
}

public class DatabaseConfigurationProvider : AsyncConfigurationProvider
{
    public override async Task LoadAync()
    {
        Data = await ...
    }
}
```

### Alternative Designs

It might be possible to add an async method into the existing type but I'm not sure how the implementation would know to use the asynchronous call.

```diff
namespace Microsoft.Extensions.Configuration

public abstract partial class ConfigurationProvider : IConfigurationProvider
{
+    public virtual Task LoadAsync();
}
```

**Comment by davidfowl** [comment]
The alternative design looks better but did you also look at all of the consumers of the Load method to understand what else would need to be changed to async to make this work in practice?

**Comment by jackbond** [comment]
Please don't stop with LoadAsync. There should also be TryGetAsync and SetAsync. Our current version of IConfigurationProvider uses Azure.Data.AppConfiguration under the hood, reads the settings into a local cache during Load, but doesn't load a secret from the underlying KeyVault until the setting is actually requested, i.e. TryGet is called. At that point an async call is made to KeyVault, but sadly GetAwaiter().GetResult() is necessary.

This is one of those instances where Microsoft libraries make it difficult to "async all the way down". In addition, with providers like Azure AppConfiguration, which can contain dozens / hundreds of secrets, you really don't want to be reading them all during Load. Ideally load would simply load the list of keys a provider contains, but not their actual values.

**Comment by davidfowl** [comment]
The current model is fundamentally incompatible with that @jackbond and if we do that then we should make a different type. The current configuration assumes you can access all of the keys in memory. What you're asking for is asynchronous access to individual keys. If we go that route then we should build a new API.

**Comment by KieranDevvs** [comment]
> The alternative design looks better but did you also look at all of the consumers of the Load method to understand what else would need to be changed to async to make this work in practice? What about the generic hosting APIs?

Ok thanks for the feedback, I'm on xmas holiday at the moment, ill revisit this in the new year.

**Comment by jackbond** [comment]
@davidfowl I'd clarify that I'm asking for asynchronous access to **values**, while the keys themselves are actually all in memory. The original poster asked for with IAsyncConfigurationSource.LoadAsync, which would meet your requirement of having all the keys in memory, and I merely took it one logical step further with TryGetAsync, which would allow the delayed loading of a value, which is actually more realistic / appropriate in distributed / microservice environments.

**Comment by davidfowl** [comment]
I don’t understand why you need asynchronous access to individual keys if they are all hydrated in memory. IMO that’s another interface and a different API completely.

**Comment by jackbond** [comment]
The keys are loaded into memory during LoadAsync(). Let's say I've got dozens of values which are stored in KeyVault via azure appconfiguration, now I've got to do dozens of KeyVault calls **during startup** to get those keys, and the reality is, any particular process (because I've got dozens of microservices) may actually only be using 1 (or even none) of them. Now let's take it one step further, what if a value changes in key vault. Want to take a guess how the appconfigurationprovider built by the azure app config team handles this? They cache the values, which is fine, and when you ask for a value, they check to see if the cached value is out of date. If it is, they trigger a call in the background to update it, BUT they return the current value.

Ultimately all this goes back to IConfiguration being designed when synchronous call to read a config file was the norm. That's obviously no longer sufficient. The original poster suggested IAsyncConfigurationSource and IAsyncConfigurationProvider, and I think everyone agrees that another interface is necessary, and my only suggestion is that, take to its logical conclusion, it ought to be IAsyncConfigurationProvider.TryGetAsync.

**Comment by davidfowl** [comment]
Right, and I don't have a problem with having a new abstraction for stores like keyvault and azure app configuration. Seems very much like https://github.com/dotnet/runtime/issues/36035.

If that's the goal of asynchrony, then I would opt for not polluting the existing API and designing a new one.

**Comment by jackbond** [comment]
Yes, and I agree with your assessment in that thread that async / pull support is necessary, whereas a separate stack dedicated to secrets is not. That thread also touched on how other elements of the stack are not async, which is really a larger issue that needs to be addressed by the c# / core teams. For example, even reading a local json config file should be an async operation, it's doing async IO after all.

**Comment by davidfowl** [comment]
>  For example, even reading a local json config file should be an async operation, it's doing async IO after all.

Loading everything up front (async or sync) is fine, and why the original API proposal (though I would change the existing interfaces and classes instead of making new ones) is easier to reason about with the existing configuration model.  

Anything that requires asynchronous access to values should be a different abstraction. So maybe this `IAsyncConfigurationProvider` proposal should be about accessing values asynchronously. Instead of loading all keys up front somehow (Do most of those systems have a cheap way to query all keys? Do we even want that?).

I think these providers would be inappropriate for things like configuration binding, which want to know all of the keys (in the entire object hierarchy) before binding. This API would be used directly in the places that need access to individual values at the moment they are queried, without caching from the source.

**Comment by jackbond** [comment]
Well I can't speak for all providers, but Azure.Data.AppConfiguration.ConfigurationClient has a method GetConfigurationSettingsAsync which allows you to read all the keys. It also loads all the values, but some of those values are merely links to the underlying KeyVault. We delay load those, and use .Result (which always bother me) in TryGet.

"So maybe this IAsyncConfigurationProvider proposal should be about accessing values asynchronously."
I think that's the critical requirement.

**Comment by davidfowl** [comment]
> Well I can't speak for all providers, but Azure.Data.AppConfiguration.ConfigurationClient has a method GetConfigurationSettingsAsync which allows you to read all the keys. It also loads all the values, but some of those values are merely links to the underlying KeyVault. We delay load those, and use .Result (which always bother me) in TryGet.

Right, to make general abstractions, we need to make sure the pattern holds. I'm making an assumption that enumerating keys without values isn't common, but if we can find places where it is, then those assumptions can change,

**Comment by jackbond** [comment]
I completely get wanting the pattern to hold broadly. But at the same time, I do miss the old Microsoft days when the attitude was, "If you don't support our pattern, it's your problem" :)

**Comment by davidfowl** [comment]
> But at the same time, I do miss the old Microsoft days when the attitude was, "If you don't support our pattern, it's your problem" :)

That's ironic because the reason the azure configuration provider and keyvault provider is not ideal today is because the existing abstraction doesn't work well with the existing model😄

**Comment by jackbond** [comment]
😄 True enough, although this feels more like an inhouse dispute, where IConfiguration simply hasn't received the modernization attention it deserves. But that's true of a lot of classes. For example, how is the FileInfo class setting its Exists property without sneaking in an async operation? ServiceCollection should support BuildServiceProviderAsync, which would allow WebApplicationBuilder to expose BuildAsync ('m guessing.) It's hard to do async all the way down when some libraries stop you dead in your tracks.

**Comment by davidfowl** [comment]
> 😄 True enough, although this feels more like an inhouse dispute, where IConfiguration simply hasn't received the modernization attention it deserves.

As one of the creators, that's not the case.

> For example, how is the FileInfo class setting its Exists property without sneaking in an async operation?

FileProviders was never really meant to exist outside of the ASP.NET Core use case of being a readonly file system for web applications. I don't think it has any of the appropriate features for a general purpose file system abstraction (as much as people want it to be).

>  ServiceCollection should support BuildServiceProviderAsync

This doesn't make any sense to me and would make everything less efficient.

**Comment by jackbond** [comment]
"As one of the creators of this abstraction, that's not the case."

To be clear, not saying it hasn't been maintained, just not "asyncified".

In regards to the FileInfo class, I'm not talking anything ASP.NET specific. System.IO.FileInfo.Exists could be potentially more than 50ms (is that still the guidance), I'd argue that it should have been Obsoleted and replaced with ExistsAsync a while back.

BuildServiceProvider is not something that you're running repeatedly in a tight loop (so less concerned about performance), but it is something that more than likely has async operations under the hood. For example, reading a connection string async, and passing that to a service like a db provider (which may actually like to initialize / warm up asynchronously.)

"Agreed. It's a pain, but everything shouldn't be async or we can just throw performance out the window."

Unless the guidance has changed, async all the way is still the official line, and being forced to use .Result due to an external API doesn't change that, it just introduces the potential for a deadlock.

As a side note, I'd like to thank you for all the work you do, and for taking time out to participate in discussions like this.

**Comment by davidfowl** [comment]
> To be clear, not saying it hasn't been maintained, just not "asyncified".

We're reluctant to do it without deeply understanding the usage. That's where you (customers) come in! 

> In regards to the FileInfo class, I'm not talking anything ASP.NET specific. System.IO.FileInfo.Exists could be potentially more than 50ms (is that still the guidance), I'd argue that it should have been Obsoleted and replaced with ExistsAsync a while back.

I don't agree. Unless we're talking about mapped network storage (cloud or not), file access is usually fast and cached at the layer. Deprecating Exists doesn't make sense IMO.

> BuildServiceProvider is not something that you're running repeatedly in a tight loop (so less concerned about performance), but it is something that more than likely has async operations under the hood. For example, reading a connection string async, and passing that to a service like a db provider (which may actually like to initialize / warm up asynchronously.)

Build doesn't do anything async, [resolution can though](https://github.com/dotnet/runtime/issues/65656).

> Unless the guidance has changed, async all the way is still the official line, and being forced to use .Result due to an external API doesn't change that, it just introduces the potential for a deadlock.

Right and everything shouldn't be async. The good news is though, making something async that was previously sync should force people to rethink all of the call sites and better understand why async is needed. The fact that it is viral is significant, and if it ends up being expensive (for example configuration binding), then we should re-think the interface and consumption.

> As a side note, I'd like to thank you for all the work you do, and for taking time out to participate in discussions like this.

Thanks 🙏🏾

**Comment by jackbond** [comment]
"Deprecating Exists doesn't make sense IMO"

Certainly not a hill I'm going to die on, and for some reason I thought they had obsoleted the sync IO stuff, so my mistake.

"Build doesn't do anything async" 

I'm delighted that you guys are looking into Asynchronous DI support because that's what I'm really after in regards to BuildServiceProviderAsync. As mentioned in that thread, async constructors are missing. Sad that they were not included with C# 5 or 6,

**Comment by KieranDevvs** [comment]
> The current model is fundamentally incompatible with that @jackbond and if we do that then we should make a different type. The current configuration assumes you can access all of the keys in memory. What you're asking for is asynchronous access to individual keys. If we go that route then we should build a new API that doesn't overlap with the existing configuration provider.

I agree with this, if you're loading the values at the time of request i.e lazy loading, then that should be a new API, or even left to the user to implement as I cant think of many other scenarios where you would also want that behavior?

> The alternative design looks better but did you also look at all of the consumers of the Load method to understand what else would need to be changed to async to make this work in practice? What about the generic hosting APIs?

The only consumer of `Load` as far as I can see, is the `Build` call. As for your comment on generic hosting API's, those API's as well as anything else that would want to leverage async configuration sources, would need to implement the feature. The approach I'm proposing is such that it would be opt-in and wouldn't change any existing behavior.

**Comment by jjxtra** [comment]
Agreed that an async interface and class should be made separately and would be used by a minority of people, but still necessary for those people.

**Comment by Rabadash8820** [comment]
Just adding a +1 here. I don't have strong feelings on whether there's a new interface or an addition to the existing `IConfigurationBuilder` interface, but some sort of async support is essential for scenarios with custom threading models. For example, I am trying to implement a [configuration provider for Unity RemoteConfig](https://github.com/DerploidEntertainment/UnityUtil/tree/main/src/UnityUtil.Configuration.RemoteConfig). RemoteConfig exposes both a synchronous `FetchConfigs` and an asynchronous `FetchConfigsAsync` method, but I can't use either:

- `Wait`ing the `Task` returned by `RemoteConfigService.FetchConfigsAsync` fails as this would lead to a deadlock with the Unity main thread that calls `IConfigurationBuilder.Build()` (as the `UnitySynchronizationContext` essentially queues tasks on that thread, rather than using the thread pool)
- `FetchConfigs` does not synchronously return settings, but invokes a `FetchCompleted` event to which calling code can subscribe, and this callback runs at an indeterminate time. My `IConfigurationSource` implementation could accept a callback and add _that_ as a listener to `FetchCompleted`, but that breaks the encapsulation/contract of configuration sources. Consumers could also wait for a short time after after building configuration to ensure that `FetchCompleted` has been triggered, but the amount of time would be arbitrary--either too short or longer than necessary.

I've [raised the latter concern](https://forum.unity.com/threads/fetchconfigs-synchronously-without-callback-to-support-ms-extensions-configuration.1481922/) on the Unity Forum, but an asynchronous version of `IConfigurationBuilder.Load` would also solve these issues, allowing calling code to `await` and play nicely with the `UnitySynchronizationContext`.

**Comment by klemmchr** [comment]
What's the status on this? Most configuration providers are asynchronous under the hood so this is desperately needed.

**Comment by KieranDevvs** [comment]
> What's the status on this? Most configuration providers are asynchronous under the hood so this is desperately needed.

You can see the status. You can always contribute to help push this along.

**Comment by robertmclaws** [comment]
@davidfowl Would be amazing if this could get into .NET 10...

**Comment by dazinator** [comment]
I worked around this by calling an InitialiseAsync() on my provider to async prefetch the configuration values, ahead of adding it and calling ConfigurationBuilder.Build() which it can then return the perfetched configuration synchronously.

However this doesn't solve the issue of `Reload` and change tokens not being asynchronous.
When a configuration provider is signalled to reload that currently is also a synchronous operation. I had to get creative and implement an Async version of a change token to solve this issue. 


@davidfowl I think this is more about the conflict of
1. We are often told that network bound io calls should be asynchronous, many libraries provide asynchronous methods to make those calls.
2. Configuration often has to be fetched from network initially.
3. The HostBuilder and configuration builder experience doesn't provide any obvious location for performing this async configuration fetching, either during the initial build, or a config reload.
4. Change tokens signal config reloads. There is a seperate item somewhere about async change tokens, I think this might help.
5. Is it OK to fetch config synchronously instead - probably yes, keeping hold of extra cycles for a smallish network call once at application startup or a few times during config reloads are runtime is most likely not a big issue. However its cognitively irksome to make network calls using synchronous api's and feels "wrong". Its also not a problem to do this asynchronously either in terms of "performance" that should not be the issue here.

**Comment by klemmchr** [comment]
To add another use case where we were facing issues:

We have a custom configuration provider in Blazor that makes an api call to get the current configuration for the web app. We are forced to do this since we're having different configurations per environment and do not want to rebuild our whole application when changing these or when rolling out our application.
When using Blazor WASM you cannot use blocking calls because it is running single threaded. Any attempt to use blocking calls (like `GetAwaiter().GetResult()` will yield a `PlatformNotSupportedException` (https://github.com/dotnet/aspnetcore/issues/57429).

Our current approach matches the one from @dazinator: we create the provider beforehand, load the config in the main entry point (where we can use async) and then add it to the configuration builder. While this approach works fine, it has some drawbacks:

1. It's clunky. We're bypassing all logic that the configuration provider has for loading configuration and we need to have quite some initialization code in our application just to support that use case.
2. We cannot use config reloading at all. That's not the worse thing you could have but it would be great if we could reload our configuration from time to time to ensure that feature flags are up-to-date. 

Besides this technical issues, this also feels weird for other reasons:

1. We were told to use async everywhere. From simple file operations to network operations, everything is async these days. It feels counter intuitive to work around this just because the config provider does not support it.
2. A lot of config providers actually need this. Even simple providers like file config providers should prefer async file APIs. Besides that, there are dozens of config providers that actually make network request that surely should not block a thread while doing this (https://github.com/Azure/AppConfiguration to name one)

Having support for an async provider should be optional, not mandatory. I don't think that introducing a breaking change in the `IConfigurationProvider` interface would be a good move, therefore I would go with a new interface `IAsyncConfigurationProvider` that could optionally inherit from `IConfigurationProvider`. Having a breaking change in the host builder would affect a lot of things already so I don't think breaking all existing configuration providers would be beneficial.

**Comment by aradalvand** [comment]
No `CancellationToken` in this proposal?
Why?

**Comment by KieranDevvs** [comment]
> No `CancellationToken` in this proposal? Why not?

It should, it was an oversight on my part. I've updated the proposal.

**Comment by KieranDevvs** [comment]
I've decided to revisit this and attempt a proof of concept but I'm struggling to come up with an API surface that doesn't have breaking changes and that isnt awful.

Currently, `ConfigurationRoot` 's constructor is the place where the `IConfiguirationProvider.Load()` is called which means further surface area would need to be changed as constructors cannot be asynchronous. An aditional method would need to be added to the `ConfigurationRoot` for it to be able to load data from both both sync and async providers.

What is the likelihood that we can have breaking changes in order to properly provide async support? I'm assuming its a near zero chance as this API is widely used?

**Comment by davidfowl** [comment]
0 chance of making a breaking change here. It need to be additive

**Comment by klemmchr** [comment]
The only way to make this non-breaking is to synchronously await Tasks in the old implementation which is a big pain point. On the other side, this is exactly how providers do it right now if they have async operations. Just keep in mind that this will cause issues on single threaded runtimes like Blazor so the host in these environments should always use the async method.

Regarding the CancellationToken: I don't see any benefit of adding them since I can't imagine a scenario where loading a configuration would be cancelled. If two config reloads are triggered at the same time the reload mechanism should rather just ignore the second one than canceling the first (so basically a stampede protection).

**Comment by KieranDevvs** [comment]
So, I got a proof of concept working which would let users opt into adding async sources and building them into a configuration via `BuildAsync()`. This would work ***reasonably*** well for users who are using `ConfigurationBuilder` directly and creating their own configuration, and would not change any behaviour in the existing APIs.

However... I would assume 80% of configurations that are built come from one of the generic hosting API's that use `IConfigurationManager`. Unfortunately, the `IConfigurationManager` is implemented in a way where it is both a configuration root and its own builder. This wouldn't be an issue by itself, but rather than relying upon the implemented `IConfigurationBuilder.Build()` method to be called in order to build itself, it loads the providers upon adding them to the config (in a non-async call) (I assume this was hacked in, in order to allow for the configuration to be used while still configuring the host.). This essentially means that adding async support to this API would only be useful to (what I assume is) a small number of people who are using `ConfigurationBuilder` directly.

At this point, I don't think this is worth adding if only a subset of people are going to benefit from it, and the configuration API would have different capabilities depending on how you use it which would just be confusing:

| API  | Async Support |
| ------------- | ------------- |
| ConfigurationBuilder.Build()  | No  |
| ConfigurationBuilder.BuildAsync() //Proposed new API | Yes  |
| WebApplicationBuilder.Build()  | No  |
| HostApplicationBuilder.Build()  | No  |
| DistributedApplicationBuilder.Build()  | No  |

**Comment by klemmchr** [comment]
Looks like the classical "this should have been async from the beginning but now the ship has sailed" scenario. That's a pity.

**Comment by aradalvand** [comment]
> Regarding the CancellationToken: I don't see any benefit of adding them since I can't imagine a scenario where loading a configuration would be cancelled.

It should be canceled when a stopping signal is sent to the application (e.g. the process is shut down).

**Comment by Rabadash8820** [comment]
Thanks for attempting a PoC @KieranDevvs!

A couple thoughts:

> It should be canceled when a stopping signal is received by the application (e.g. the process is shut down) — a la `stoppingToken` in `BackgroundService.ExecuteAsync`.

@aradalvand Agreed. Especially since we're talking about this new API being used mostly outside of host builders, where the `BuildAsync()` call would be explicit.

> At this point, I don't think this is worth adding if only a subset of people are going to benefit from it, and the configuration API would have different capabilities depending on how you use it which would just be confusing:

@KieranDevvs I cannot speak to the number of developers using `IConfigurationBuilder` outside of hosts either, but I can definitely think of some .NET environments where one might want to use the MsEx configuration libraries but not have access to a host builder:
- Adding configuration to legacy apps that aren't using a host builder
- Using the `IConfigurationBuilders` purely for file parsing (we are currently doing this at my job to parse INI files!)
- Adding configuration to Unity projects, as I [described above](https://github.com/dotnet/runtime/issues/79193#issuecomment-1685439773). Again, I can't speak to how many developers are actually doing this in Unity, but I know I am. 🤷‍♂ Unity is probably one of the larger use cases for MsEx libraries where a host builder is unavailable (along with code in other game engines like Godot).
- Potentially Xamarin projects also? I know they used to be built with Mono, like Unity, but idk if modern Xamarin projects have a `MobileApplicationBuilder` or similar. Even if they do, my legacy app argument applies here as well. 👌

**Comment by robertmclaws** [comment]
Gonna throw this out there: in .NET 10 would it be possible to 

- create the future for this API with the new Async pattern,
- have an Sync-over-Async method that triggers any existing sync registrations for compatibility (which could potentially be called automatically in the Async paths),
- and puts the old codepaths on a path to deprecation?

Since loading configuration almost always involves file / network IO and this should have been Async from the start (but it likely wasn't because Program.Main was only made async recently and this pattern is much older), it should be possible to transition the ecosystem to a new world where the entire HostBuilder codepath has Async sypport.

**Comment by klemmchr** [comment]
This would still be a breaking change at some point later on. I'm pretty sure this is also covered by the guideline "no breaking change".

**Comment by dazinator** [comment]
How about this path

1. Introduce a new `ConfigurationBuilderAsync` alternative to the current subsystem. It can still implement the same [IConfigurationBuilder](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.iconfigurationbuilder?view=net-9.0-pp) interface, but it could also be extended with async variants.
2. A path where users can explicitly use the async variant instead of the sync one, with the host.
3. The host detects if the async one is in use, and uses its async methods rather than the sync methods.

The goal here is to think of a way to allow an async alternative builder / provider to be defined and swapped into use for those who need it.


The async builder, could work with a new (extended?) async provider definition.

**Comment by KieranDevvs** [comment]
> How about this path
> 
> 1. Introduce a new `AsyncConfigurationBuilder` alternative to the current subsystem. It can still implement the same [IConfigurationBuilder](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.iconfigurationbuilder?view=net-9.0-pp) interface, but it could also be extended with the necessary async method variants.
> 2. A path where users can explicitly use the async variant instead of the sync one, with the host.
> 3. The host detects if the async one is in use, and uses its async methods rather than the sync methods.
> 
> The goal here is to think of a way to allow an async alternative builder / provider to be defined and swapped into use for those who need it.
> 
> The async builder, could work with a new (extended?) async provider definition.

The issue is point 3.
We need to add `Task` based methods to the `IConfigurationManager` but doing so would be a breaking change.
So the only available option is to create a superseding interface that implements `IConfigurationManager` and adds the async based methods. Lets call this interface `IAsyncConfigurationManager`.
We now want to be able to use this in the generic host APIs. If we take a look at `IHostApplicationBuilder` we can see that it defines the configuration via a property of the type `IConfigurationManager`. We cant change or add to this interface as it would also be a breaking change.

TLDR;
1) Cant change or add to: `IConfigurationManager`, can only inherit and extend.
2) Generic host API's such as `IHostApplicationBuilder` expose their configuration building via `IConfigurationManager`.
3) The generic host APIs cant be changed or added to, can only inherit and extend.

**Comment by robertmclaws** [comment]
Sounds like we just need to introduce an AsyncHost and start over all the way down the Host stack.

Which is probably a good idea, because it's really dumb that HostBuilder and WebHostBuilder don't implement the same interface.

**Comment by davidfowl** [comment]
WebHostBuilder is not getting updates, WebApplicationBuilder is what we’ve been evolving, similarly Host.CreateApplicationBuilder

**Comment by KieranDevvs** [comment]
Id like feedback on this idea:

We move ahead with `IAsyncConfigurationBuilder` and the other async interface types for the direct usage of configuration builder and the ability to register, and asynchronously load providers.

We also leave the generic host API's as they are (exposing IConfigurationBuilder).

Use case 1 - Using ConfigurationBuilder directly:
Users now instanciate AsyncConfigurationBuilder and load their synchronous and asynchronous providers as expected and build returns an `IAsyncConfigurationRoot` (an extension of `ConfigurationRoot` with the ability to asyncronously reload the data soruces).

Use case 2 - Using one of the generic host API's:
Users instanciate the relevant host builder.
They can either call `.Configuration.Add(...)` to add their syncrnonous provider types, or they can cast the host configuration manager to `IAsyncConfigurationManager` to access the relevant async API's

This is what I have in my current PoC:

![Image](https://github.com/user-attachments/assets/6df71490-7282-4a4c-a190-97719151fff0)

Granted, I'm fully aware that casting the `ConfigurationManager` isnt discoverable, expected or a nice user experience, but it does solve all of the breaking change issues without having to provide a set of new generic host API's.
I'm hoping that seeing as its more of an advanced use case, that if its documented well, it will be sufficient?

**Comment by Rabadash8820** [comment]
Some thoughts/questions:

1. Instead of casting `ApplicationHistBuilder.Cinfiguration`, could we add an `IAsyncConfigurationBuilder AsyncConfiguration` property? The two properties could even point to the same instance potentially. Still a slightly awkward API, but much more discoverable than casting IMO.
2. Does `AsyncConfigurationBuilder.AddAsync()` really need to be async? I thought all I/O happened when you actually _build_ the builder; and adding a configuration source would basically just synchronously add it to some private collection.

**Comment by KieranDevvs** [comment]
> Some thoughts/questions:
> 
> 1. Instead of casting `ApplicationHistBuilder.Cinfiguration`, could we add an `IAsyncConfigurationBuilder AsyncConfiguration` property? The two properties could even point to the same instance potentially. Still a slightly awkward API, but much more discoverable than casting IMO.
> 2. Does `AsyncConfigurationBuilder.AddAsync()` really need to be async? I thought all I/O happened when you actually _build_ the builder; and adding a configuration source would basically just synchronously add it to some private collection.

1) I'm happy to have it as an option and see what others think, I don't feel strongly either way but I think I would prefer the cast as A duplicate property would be confusing. I would assume that the two properties were isolated from each other.
2) Unfortunately yes. Here's why:
    - When using the `AsyncConfigurationBuilder` directly, its not needed as the data is loaded upon the `BuildAsync()` call.
    - When using `ConfigurationManager` (the type that the generic host API's expose), the manager itself is its own builder and configuration root. This means that when you add a data source, it immediate loads the data upon the `.Add/.AddAsync` call. This is done to let the user access the configuration values when configuring other parts of the host before it is built (think DI etc).

**Comment by KieranDevvs** [comment]
> > It means that anyone who has their own type that implements the interface now has to implement said property
> 
> Not necessarily, if we use a [default property implementation](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods#detailed-design) on the interface. Though idk if the implementation would really be simple enough to be "default". Your extension method idea would work too.
> 
> > the manager itself is its own builder and configuration root. This means that when you add a data source, it immediate loads the data
> 
> I see, thanks for clarifying. I suppose `ConfigurationManager.Add()` could be modified to call `Add()` and then wait on `BuildAsync()`, rather than waiting on an `.AddAsync()` call. This would not be a breaking API change, just a change to internal logic. Though that would mean potentially repeated calls to `AsyncConfigurationBuilder.BuildAsync()` and idk if that's allowed.

Note that in my PoC, I've also added `IAsyncConfigurationSource` to allow for asynchronous construction of the provider. The await not only loads the data from the source, but it also constructs the source provider asynchronously and adds it to the managers internal sources. Part of this feature extension is to allow users to implement the interface on their own types and asynchronously construct a source if needed.

```cs
        private async Task AddSourceAsync(IAsyncConfigurationSource source)
        {
            IAsyncConfigurationProvider provider = await source.BuildAsync(this).ConfigureAwait(false);

            provider.Load();
            await provider.LoadAsync().ConfigureAwait(false);
            _changeTokenRegistrations.Add(ChangeToken.OnChange(provider.GetReloadToken, RaiseChanged));

            _providerManager.AddProvider(provider);
            RaiseChanged();
        }
```

If we forget about `IAsyncConfigurationSource` for a second. You mean it would look something like this?

```cs
        HostApplicationBuilder hostBuilder = Host.CreateApplicationBuilder(args);
        hostBuilder.Configuration.Add(new MyAsyncConfigurationSource());

        await hostBuilder.Configuration.BuildAsync(); //Build before usage.
        hostBuilder.Services.Configure<MyType>(() => hostBuilder.Configuration...); //Usage

        var host = hostBuilder.Build();
```

If so, I think that would lead to confusion as adding an `IConfigurationSource` would result in available access straight away, but adding an `IAsyncConfigurationSource` would mean your data was missing until you called `BuildAsync` on the configuration.

**Comment by KieranDevvs** [comment]
I've updated the proposal to reflect what has been discussed so far.

**Comment by robertmclaws** [comment]
> WebHostBuilder is not getting updates, WebApplicationBuilder is what we’ve been evolving, similarly Host.CreateApplicationBuilder (those implement the same interface )

I will look into this, thanks @davidfowl! Maybe my Host-based extensions have all been supporting the wrong interfaces.

**Comment by KieranDevvs** [comment]
If no one has any comments, can this be marked as ready for review?
